# 20. 多任務學習 (MTL) vs 單任務學習 (STL) 比較

## 目的
比較多任務學習（MTL）與單任務學習（STL）在三高疾病預測上的表現。

## 假說
MTL 可能透過學習跨任務的共享表徵來提升預測效能。

## 背景
根據共病分析：
- 三高疾病之間的直接相關性較弱（Phi < 0.1）
- 高血壓 ↔ 高血糖：OR=1.72（中度關聯）
- 高血壓 ↔ 高血脂：OR=1.07（近乎獨立）

## 日期：2026-01-13

In [1]:
# 安裝 PyTorch（若未安裝）
try:
    import torch
    print(f"PyTorch 已安裝：{torch.__version__}")
except ImportError:
    print("正在安裝 PyTorch...")
    !pip install torch
    print("PyTorch 安裝完成")

PyTorch 已安裝：1.13.1+cpu


In [2]:
# 匯入套件
import pandas as pd
import numpy as np
import time
import warnings
warnings.filterwarnings('ignore')

from sklearn.model_selection import StratifiedGroupKFold
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.multioutput import MultiOutputClassifier
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import roc_auc_score, f1_score, confusion_matrix
from sklearn.base import clone

import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import DataLoader, TensorDataset

print("套件載入完成")
print(f"PyTorch 版本：{torch.__version__}")

套件載入完成
PyTorch 版本：1.13.1+cpu


## 1. 載入資料

In [3]:
# 載入滑動視窗資料
df = pd.read_csv('../../data/01_primary/SUA/processed/SUA_sliding_window.csv')
print(f"資料：{len(df):,} 筆樣本，{df['patient_id'].nunique():,} 位患者")

# 特徵
feature_cols = [
    'sex', 'Age',
    'FBG_Tinput1', 'TC_Tinput1', 'Cr_Tinput1', 'UA_Tinput1', 'GFR_Tinput1', 'BMI_Tinput1', 'SBP_Tinput1', 'DBP_Tinput1',
    'FBG_Tinput2', 'TC_Tinput2', 'Cr_Tinput2', 'UA_Tinput2', 'GFR_Tinput2', 'BMI_Tinput2', 'SBP_Tinput2', 'DBP_Tinput2',
    'Delta_FBG', 'Delta_TC', 'Delta_Cr', 'Delta_UA', 'Delta_GFR', 'Delta_BMI', 'Delta_SBP', 'Delta_DBP'
]

X = df[feature_cols]
groups = df['patient_id']

# 目標變數（將 1/2 轉換為 0/1）
y_htn = (df['hypertension_target'] == 2).astype(int)
y_hg = (df['hyperglycemia_target'] == 2).astype(int)
y_dl = (df['dyslipidemia_target'] == 2).astype(int)

# MTL 用的組合目標矩陣
Y = pd.DataFrame({
    'HTN': y_htn,
    'HG': y_hg,
    'DL': y_dl
})

print(f"\n特徵數：{len(feature_cols)}")
print(f"任務：{list(Y.columns)}")
print(f"\n盛行率：")
for col in Y.columns:
    print(f"  {col}：{Y[col].mean()*100:.1f}%")

資料：13,514 筆樣本，6,056 位患者

特徵數：26
任務：['HTN', 'HG', 'DL']

盛行率：
  HTN：19.3%
  HG：5.9%
  DL：7.9%


## 2. 定義模型

### STL（單任務學習）
- 3 個獨立模型，各預測一種疾病

### MTL（多任務學習）
- 1 個共享模型，同時預測 3 種疾病

In [4]:
# PyTorch MTL 模型
class MTLNetwork(nn.Module):
    """多任務學習網路（含共享層）"""
    def __init__(self, input_dim, hidden_dims=[64, 32], n_tasks=3):
        super().__init__()
        
        # 共享層
        layers = []
        prev_dim = input_dim
        for h_dim in hidden_dims:
            layers.extend([
                nn.Linear(prev_dim, h_dim),
                nn.ReLU(),
                nn.Dropout(0.2)
            ])
            prev_dim = h_dim
        self.shared = nn.Sequential(*layers)
        
        # 任務專屬輸出層
        self.heads = nn.ModuleList([
            nn.Linear(hidden_dims[-1], 1) for _ in range(n_tasks)
        ])
    
    def forward(self, x):
        shared_repr = self.shared(x)
        outputs = [torch.sigmoid(head(shared_repr)) for head in self.heads]
        return torch.cat(outputs, dim=1)


class STLNetwork(nn.Module):
    """單任務學習網路（相同架構，但獨立訓練）"""
    def __init__(self, input_dim, hidden_dims=[64, 32]):
        super().__init__()
        
        layers = []
        prev_dim = input_dim
        for h_dim in hidden_dims:
            layers.extend([
                nn.Linear(prev_dim, h_dim),
                nn.ReLU(),
                nn.Dropout(0.2)
            ])
            prev_dim = h_dim
        layers.append(nn.Linear(hidden_dims[-1], 1))
        
        self.network = nn.Sequential(*layers)
    
    def forward(self, x):
        return torch.sigmoid(self.network(x))

print("模型定義完成")

模型定義完成


In [5]:
def train_mtl_model(X_train, Y_train, X_val, Y_val, epochs=100, lr=0.001, batch_size=256):
    """訓練 MTL 模型"""
    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
    
    # 準備資料
    X_train_t = torch.FloatTensor(X_train).to(device)
    Y_train_t = torch.FloatTensor(Y_train.values).to(device)
    X_val_t = torch.FloatTensor(X_val).to(device)
    Y_val_t = torch.FloatTensor(Y_val.values).to(device)
    
    train_dataset = TensorDataset(X_train_t, Y_train_t)
    train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)
    
    # 模型
    model = MTLNetwork(X_train.shape[1]).to(device)
    
    # 類別權重（處理不平衡資料）
    pos_weights = []
    for i in range(Y_train.shape[1]):
        pos_rate = Y_train.iloc[:, i].mean()
        pos_weights.append((1 - pos_rate) / pos_rate)
    pos_weight = torch.FloatTensor(pos_weights).to(device)
    
    criterion = nn.BCELoss(reduction='none')
    optimizer = optim.Adam(model.parameters(), lr=lr)
    
    # 訓練
    best_auc = 0
    patience = 10
    patience_counter = 0
    
    for epoch in range(epochs):
        model.train()
        for X_batch, Y_batch in train_loader:
            optimizer.zero_grad()
            outputs = model(X_batch)
            
            # 加權損失
            loss = criterion(outputs, Y_batch)
            weights = torch.where(Y_batch == 1, pos_weight, torch.ones_like(Y_batch))
            loss = (loss * weights).mean()
            
            loss.backward()
            optimizer.step()
        
        # 驗證
        model.eval()
        with torch.no_grad():
            val_outputs = model(X_val_t).cpu().numpy()
            val_aucs = [roc_auc_score(Y_val.iloc[:, i], val_outputs[:, i]) for i in range(3)]
            mean_auc = np.mean(val_aucs)
            
            if mean_auc > best_auc:
                best_auc = mean_auc
                patience_counter = 0
            else:
                patience_counter += 1
                if patience_counter >= patience:
                    break
    
    # 最終預測
    model.eval()
    with torch.no_grad():
        predictions = model(X_val_t).cpu().numpy()
    
    return predictions, model


def train_stl_models(X_train, Y_train, X_val, Y_val, epochs=100, lr=0.001, batch_size=256):
    """訓練 3 個獨立的 STL 模型"""
    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
    
    X_train_t = torch.FloatTensor(X_train).to(device)
    X_val_t = torch.FloatTensor(X_val).to(device)
    
    all_predictions = []
    models = []
    
    for task_idx in range(Y_train.shape[1]):
        y_train = Y_train.iloc[:, task_idx].values
        y_val = Y_val.iloc[:, task_idx].values
        
        y_train_t = torch.FloatTensor(y_train).unsqueeze(1).to(device)
        
        train_dataset = TensorDataset(X_train_t, y_train_t)
        train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)
        
        model = STLNetwork(X_train.shape[1]).to(device)
        
        # 類別權重
        pos_rate = y_train.mean()
        pos_weight = torch.FloatTensor([(1 - pos_rate) / pos_rate]).to(device)
        
        criterion = nn.BCELoss(reduction='none')
        optimizer = optim.Adam(model.parameters(), lr=lr)
        
        best_auc = 0
        patience = 10
        patience_counter = 0
        
        for epoch in range(epochs):
            model.train()
            for X_batch, y_batch in train_loader:
                optimizer.zero_grad()
                outputs = model(X_batch)
                
                loss = criterion(outputs, y_batch)
                weights = torch.where(y_batch == 1, pos_weight, torch.ones_like(y_batch))
                loss = (loss * weights).mean()
                
                loss.backward()
                optimizer.step()
            
            model.eval()
            with torch.no_grad():
                val_pred = model(X_val_t).cpu().numpy().flatten()
                val_auc = roc_auc_score(y_val, val_pred)
                
                if val_auc > best_auc:
                    best_auc = val_auc
                    patience_counter = 0
                else:
                    patience_counter += 1
                    if patience_counter >= patience:
                        break
        
        model.eval()
        with torch.no_grad():
            predictions = model(X_val_t).cpu().numpy().flatten()
        
        all_predictions.append(predictions)
        models.append(model)
    
    return np.column_stack(all_predictions), models

print("訓練函式定義完成")

訓練函式定義完成


## 3. 執行 5-Fold 交叉驗證實驗

In [6]:
# 5-Fold CV（GroupKFold）— 含完整分類指標
def compute_metrics(y_true, y_prob, threshold=0.5):
    """計算 AUC, Sensitivity, Specificity, F1"""
    auc = roc_auc_score(y_true, y_prob)
    y_pred = (y_prob >= threshold).astype(int)
    tn, fp, fn, tp = confusion_matrix(y_true, y_pred).ravel()
    sens = tp / (tp + fn) if (tp + fn) > 0 else 0
    spec = tn / (tn + fp) if (tn + fp) > 0 else 0
    f1 = f1_score(y_true, y_pred)
    return auc, sens, spec, f1

n_splits = 5
cv = StratifiedGroupKFold(n_splits=n_splits, shuffle=True, random_state=42)

metrics_keys = ['HTN', 'HG', 'DL', 'time',
                'HTN_sens', 'HG_sens', 'DL_sens',
                'HTN_spec', 'HG_spec', 'DL_spec',
                'HTN_f1', 'HG_f1', 'DL_f1']
results = {
    'MTL': {k: [] for k in metrics_keys},
    'STL': {k: [] for k in metrics_keys}
}

print("=" * 70)
print("5-Fold CV：MTL vs STL（含 Sensitivity / Specificity / F1）")
print("=" * 70)

for fold, (train_idx, test_idx) in enumerate(cv.split(X, y_htn, groups)):
    print(f"\nFold {fold+1}/{n_splits}")
    
    X_train, X_test = X.iloc[train_idx], X.iloc[test_idx]
    Y_train, Y_test = Y.iloc[train_idx], Y.iloc[test_idx]
    
    train_patients = set(groups.iloc[train_idx])
    test_patients = set(groups.iloc[test_idx])
    assert len(train_patients & test_patients) == 0
    
    scaler = StandardScaler()
    X_train_scaled = scaler.fit_transform(X_train)
    X_test_scaled = scaler.transform(X_test)
    
    # --- MTL ---
    start_time = time.time()
    mtl_preds, _ = train_mtl_model(X_train_scaled, Y_train, X_test_scaled, Y_test)
    mtl_time = time.time() - start_time
    results['MTL']['time'].append(mtl_time)
    
    for i, task in enumerate(['HTN', 'HG', 'DL']):
        auc, sens, spec, f1 = compute_metrics(Y_test.iloc[:, i].values, mtl_preds[:, i])
        results['MTL'][task].append(auc)
        results['MTL'][f'{task}_sens'].append(sens)
        results['MTL'][f'{task}_spec'].append(spec)
        results['MTL'][f'{task}_f1'].append(f1)
    
    print(f"  MTL: HTN={results['MTL']['HTN'][-1]:.3f}, HG={results['MTL']['HG'][-1]:.3f}, DL={results['MTL']['DL'][-1]:.3f} ({mtl_time:.1f}s)")
    
    # --- STL ---
    start_time = time.time()
    stl_preds, _ = train_stl_models(X_train_scaled, Y_train, X_test_scaled, Y_test)
    stl_time = time.time() - start_time
    results['STL']['time'].append(stl_time)
    
    for i, task in enumerate(['HTN', 'HG', 'DL']):
        auc, sens, spec, f1 = compute_metrics(Y_test.iloc[:, i].values, stl_preds[:, i])
        results['STL'][task].append(auc)
        results['STL'][f'{task}_sens'].append(sens)
        results['STL'][f'{task}_spec'].append(spec)
        results['STL'][f'{task}_f1'].append(f1)
    
    print(f"  STL: HTN={results['STL']['HTN'][-1]:.3f}, HG={results['STL']['HG'][-1]:.3f}, DL={results['STL']['DL'][-1]:.3f} ({stl_time:.1f}s)")

print("\n" + "=" * 70)
print("交叉驗證完成")
print("=" * 70)

5-Fold CV：MTL vs STL（含 Sensitivity / Specificity / F1）

Fold 1/5
  MTL: HTN=0.744, HG=0.921, DL=0.880 (16.7s)
  STL: HTN=0.758, HG=0.923, DL=0.880 (18.5s)

Fold 2/5
  MTL: HTN=0.731, HG=0.938, DL=0.865 (4.4s)
  STL: HTN=0.742, HG=0.937, DL=0.866 (12.2s)

Fold 3/5
  MTL: HTN=0.744, HG=0.927, DL=0.857 (12.7s)
  STL: HTN=0.752, HG=0.929, DL=0.858 (19.0s)

Fold 4/5
  MTL: HTN=0.699, HG=0.936, DL=0.865 (11.0s)
  STL: HTN=0.709, HG=0.936, DL=0.864 (14.9s)

Fold 5/5
  MTL: HTN=0.747, HG=0.941, DL=0.873 (10.3s)
  STL: HTN=0.750, HG=0.940, DL=0.876 (14.9s)

交叉驗證完成


## 4. 結果摘要

In [7]:
# 完整摘要（AUC + Sensitivity + Specificity + F1 + 訓練時間 + 參數量）
print("=" * 70)
print("結果：MTL vs STL 完整比較（5-Fold CV）")
print("=" * 70)

# --- 參數量計算 ---
input_dim = len(feature_cols)  # 26
mtl_model = MTLNetwork(input_dim)
stl_model = STLNetwork(input_dim)
mtl_params = sum(p.numel() for p in mtl_model.parameters())
stl_params = sum(p.numel() for p in stl_model.parameters())

print(f"\n【參數量】")
print(f"  MTL（1 個共享網路 + 3 heads）：{mtl_params:,} 參數")
print(f"  STL（3 個獨立網路）：{stl_params:,} × 3 = {stl_params*3:,} 參數")
print(f"  MTL / STL 比例：{mtl_params / (stl_params*3):.1%}")

# --- 訓練時間 ---
print(f"\n【訓練時間】")
print(f"  MTL：{np.mean(results['MTL']['time']):.1f} ± {np.std(results['MTL']['time']):.1f} s/fold")
print(f"  STL：{np.mean(results['STL']['time']):.1f} ± {np.std(results['STL']['time']):.1f} s/fold")
ratio = np.mean(results['STL']['time']) / np.mean(results['MTL']['time'])
print(f"  STL / MTL 時間比：{ratio:.2f}x")

# --- AUC 比較（原格式） ---
print(f"\n【AUC 比較】")
summary = []
for method in ['MTL', 'STL']:
    for task in ['HTN', 'HG', 'DL']:
        aucs = results[method][task]
        summary.append({
            '方法': method,
            '任務': task,
            'AUC_mean': np.mean(aucs),
            'AUC_std': np.std(aucs)
        })
summary_df = pd.DataFrame(summary)
print(summary_df.pivot(index='任務', columns='方法', values='AUC_mean').round(3))

# --- 各指標比較表 ---
print(f"\n【各指標比較】")
print(f"\n| 疾病 | 指標 | MTL | STL | 差異 |")
print(f"|------|------|:---:|:---:|:----:|")
for task in ['HTN', 'HG', 'DL']:
    for metric, key in [('AUC', task), ('Sens', f'{task}_sens'), ('Spec', f'{task}_spec'), ('F1', f'{task}_f1')]:
        mtl_val = np.mean(results['MTL'][key])
        stl_val = np.mean(results['STL'][key])
        diff = mtl_val - stl_val
        print(f"| {task} | {metric} | {mtl_val:.3f} | {stl_val:.3f} | {diff:+.3f} |")

# --- 論文用表格格式 ---
print(f"\n\n{'='*70}")
print(f"【論文用格式（表 6-15 擴充）】")
print(f"{'='*70}")
print(f"\n| 指標 | 方法 | 高血壓 | 高血糖 | 高血脂 |")
print(f"|------|------|:------:|:------:|:------:|")
for metric_name, keys in [('AUC', ['HTN', 'HG', 'DL']),
                            ('Sensitivity', ['HTN_sens', 'HG_sens', 'DL_sens']),
                            ('Specificity', ['HTN_spec', 'HG_spec', 'DL_spec']),
                            ('F1', ['HTN_f1', 'HG_f1', 'DL_f1'])]:
    for method in ['MTL', 'STL']:
        vals = [f"{np.mean(results[method][k]):.3f}" for k in keys]
        print(f"| {metric_name} | {method} | {vals[0]} | {vals[1]} | {vals[2]} |")
    diffs = [f"{np.mean(results['MTL'][k]) - np.mean(results['STL'][k]):+.3f}" for k in keys]
    print(f"| {metric_name} | Δ(MTL−STL) | {diffs[0]} | {diffs[1]} | {diffs[2]} |")

print(f"\n訓練時間：MTL {np.mean(results['MTL']['time']):.1f}s vs STL {np.mean(results['STL']['time']):.1f}s (ratio {ratio:.2f}x)")
print(f"參數量：MTL {mtl_params:,} vs STL {stl_params*3:,} (ratio {mtl_params/(stl_params*3):.1%})")

結果：MTL vs STL 完整比較（5-Fold CV）

【參數量】
  MTL（1 個共享網路 + 3 heads）：3,907 參數
  STL（3 個獨立網路）：3,841 × 3 = 11,523 參數
  MTL / STL 比例：33.9%

【訓練時間】
  MTL：11.0 ± 4.0 s/fold
  STL：15.9 ± 2.5 s/fold
  STL / MTL 時間比：1.44x

【AUC 比較】
方法     MTL    STL
任務               
DL   0.868  0.869
HG   0.932  0.933
HTN  0.733  0.742

【各指標比較】

| 疾病 | 指標 | MTL | STL | 差異 |
|------|------|:---:|:---:|:----:|
| HTN | AUC | 0.733 | 0.742 | -0.009 |
| HTN | Sens | 0.797 | 0.798 | -0.001 |
| HTN | Spec | 0.564 | 0.581 | -0.017 |
| HTN | F1 | 0.440 | 0.450 | -0.010 |
| HG | AUC | 0.932 | 0.933 | -0.001 |
| HG | Sens | 0.860 | 0.874 | -0.015 |
| HG | Spec | 0.870 | 0.870 | -0.000 |
| HG | F1 | 0.439 | 0.445 | -0.006 |
| DL | AUC | 0.868 | 0.869 | -0.001 |
| DL | Sens | 0.828 | 0.819 | +0.009 |
| DL | Spec | 0.757 | 0.765 | -0.008 |
| DL | F1 | 0.356 | 0.360 | -0.004 |


【論文用格式（表 6-15 擴充）】

| 指標 | 方法 | 高血壓 | 高血糖 | 高血脂 |
|------|------|:------:|:------:|:------:|
| AUC | MTL | 0.733 | 0.932 | 0.868 |
| AUC | STL | 0.742 | 0

In [8]:
# 詳細比較
print("\n" + "=" * 70)
print("詳細比較")
print("=" * 70)

print("\n| 任務 | MTL AUC | STL AUC | 差異 | 勝出 |")
print("|------|---------|---------|------|------|")

for task in ['HTN', 'HG', 'DL']:
    mtl_auc = np.mean(results['MTL'][task])
    stl_auc = np.mean(results['STL'][task])
    diff = mtl_auc - stl_auc
    winner = 'MTL' if diff > 0.005 else 'STL' if diff < -0.005 else '持平'
    print(f"| {task} | {mtl_auc:.3f} | {stl_auc:.3f} | {diff:+.3f} | {winner} |")

# 整體
mtl_avg = np.mean([np.mean(results['MTL'][t]) for t in ['HTN', 'HG', 'DL']])
stl_avg = np.mean([np.mean(results['STL'][t]) for t in ['HTN', 'HG', 'DL']])
print(f"| **平均** | {mtl_avg:.3f} | {stl_avg:.3f} | {mtl_avg-stl_avg:+.3f} | {'MTL' if mtl_avg > stl_avg else 'STL'} |")


詳細比較

| 任務 | MTL AUC | STL AUC | 差異 | 勝出 |
|------|---------|---------|------|------|
| HTN | 0.733 | 0.742 | -0.009 | STL |
| HG | 0.932 | 0.933 | -0.001 | 持平 |
| DL | 0.868 | 0.869 | -0.001 | 持平 |
| **平均** | 0.844 | 0.848 | -0.004 | STL |


In [9]:
# 儲存結果
results_df = pd.DataFrame([
    {'方法': method, '任務': task, 'Fold': fold+1, 'AUC': auc}
    for method in ['MTL', 'STL']
    for task in ['HTN', 'HG', 'DL']
    for fold, auc in enumerate(results[method][task])
])
results_df.to_csv('../../results/mtl_vs_stl_results.csv', index=False)
print("已儲存：results/mtl_vs_stl_results.csv")

已儲存：results/mtl_vs_stl_results.csv


## 5. 結論

### 主要發現

1. **預測效能**：
   - MTL 與 STL 差異約 ~0.3%，無顯著差異

2. **訓練時間**：
   - MTL：較快（共享計算）
   - STL：較慢（3 個獨立模型）

3. **MTL 未能超越 STL 的可能原因**：
   - 三高疾病之間的相關性弱（Phi < 0.1）
   - 樣本數（13,514）對 STL 而言已足夠
   - 各任務難度差異大（HG AUC=0.94 vs HTN AUC=0.74）

### 結論

在本資料集上，STL 與 MTL 表現相近。
當以下情況成立時，MTL 的優勢（共享表徵）有限：
- 任務之間相關性弱
- 資料量足以進行獨立學習